File: java/src/main/java/com/sketches_oxide/membership/RibbonFilter.java
package com.sketches_oxide.membership;

import com.sketches_oxide.NativeSketch;
import com.sketches_oxide.MergeableSketch;
import com.sketches_oxide.native.SketchOxideNative;

/**
 * RibbonFilter - Space-Optimal Membership Filter with Build Phase.
 *
 * A highly space-efficient filter based on ribbon (Rapid Incremental Boolean
 * Banded matrix AND target) construction. Achieves approximately 7 bits per key
 * at 1% FPR, making it one of the most space-efficient practical filters.
 *
 * <p><strong>Algorithm:</strong>
 * <ul>
 *   <li>Uses Gaussian elimination on a banded matrix</li>
 *   <li>Items are inserted, then filter is "built" (finalized)</li>
 *   <li>After building, queries are very fast</li>
 *   <li>Achieves near-information-theoretic space lower bound</li>
 * </ul>
 *
 * <p><strong>Two-Phase Usage:</strong>
 * <ol>
 *   <li><strong>Insert Phase:</strong> Add all items using {@link #insert(byte[])}</li>
 *   <li><strong>Build Phase:</strong> Call {@link #build()} to finalize the filter</li>
 *   <li><strong>Query Phase:</strong> Use {@link #contains(byte[])} for lookups</li>
 * </ol>
 *
 * <p><strong>Use cases:</strong>
 * <ul>
 *   <li>Memory-constrained environments</li>
 *   <li>Large-scale key-value stores</li>
 *   <li>Database index structures (LSM trees)</li>
 *   <li>Any application prioritizing space efficiency</li>
 * </ul>
 *
 * <p><strong>Space Efficiency:</strong>
 * <ul>
 *   <li>~7 bits/key at 1% FPR (vs ~10 for Bloom)</li>
 *   <li>~10 bits/key at 0.1% FPR (vs ~14 for Bloom)</li>
 *   <li>~14 bits/key at 0.01% FPR (vs ~20 for Bloom)</li>
 * </ul>
 *
 * <p><strong>Important:</strong> You must call {@link #build()} after inserting all
 * items and before querying. Queries before build() will return incorrect results.
 *
 * <p><strong>Example:</strong>
 * <pre>
 * try (RibbonFilter rf = new RibbonFilter(100000, 0.01)) {
 *     // Phase 1: Insert all items
 *     for (String key : allKeys) {
 *         rf.insert(key.getBytes());
 *     }
 *
 *     // Phase 2: Build the filter (REQUIRED)
 *     rf.build();
 *
 *     // Phase 3: Query
 *     if (rf.contains("some-key".getBytes())) {
 *         System.out.println("Key might exist");
 *     }
 * }
 * </pre>
 *
 * @see BloomFilter
 * @see BinaryFuseFilter
 * @see <a href="https://arxiv.org/abs/2103.02515">Ribbon Filter Paper</a>
 */
public final class RibbonFilter extends NativeSketch implements MergeableSketch<RibbonFilter> {

    private final long n;
    private final double fpr;
    private boolean built = false;

    /**
     * Create a new RibbonFilter.
     *
     * @param n expected number of items to insert
     * @param fpr desired false positive rate (0 &lt; fpr &lt; 1)
     * @throws IllegalArgumentException if n &lt;= 0 or fpr not in (0, 1)
     * @throws RuntimeException if native memory allocation fails
     */
    public RibbonFilter(long n, double fpr) {
        if (n <= 0) {
            throw new IllegalArgumentException("n must be positive, got: " + n);
        }
        if (fpr <= 0 || fpr >= 1) {
            throw new IllegalArgumentException("fpr must be in range (0, 1), got: " + fpr);
        }

        this.n = n;
        this.fpr = fpr;
        this.nativePtr = SketchOxideNative.ribbonfilter_new(n, fpr);

        if (this.nativePtr == 0) {
            throw new RuntimeException("Failed to allocate RibbonFilter");
        }
    }

    /**
     * Private constructor for deserialization.
     */
    private RibbonFilter(long nativePtr, long n, double fpr, boolean built) {
        this.nativePtr = nativePtr;
        this.n = n;
        this.fpr = fpr;
        this.built = built;
    }

    /**
     * Insert an item into the filter.
     *
     * Items can only be inserted before {@link #build()} is called.
     *
     * @param item the item to insert (as raw bytes)
     * @throws IllegalStateException if the filter has been closed or already built
     * @throws NullPointerException if item is null
     */
    public void insert(byte[] item) {
        checkAlive();
        if (built) {
            throw new IllegalStateException("Cannot insert after build() has been called");
        }
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        SketchOxideNative.ribbonfilter_insert(nativePtr, item);
    }

    /**
     * Insert a string item into the filter.
     *
     * @param item the string to insert
     * @throws IllegalStateException if the filter has been closed or already built
     * @throws NullPointerException if item is null
     */
    public void insert(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        insert(item.getBytes());
    }

    /**
     * Build (finalize) the filter.
     *
     * This method must be called after inserting all items and before querying.
     * After calling build(), no more items can be inserted, but the filter
     * becomes queryable.
     *
     * <p><strong>Note:</strong> Building is a computationally intensive operation
     * that performs Gaussian elimination. For very large filters, this may take
     * noticeable time.
     *
     * @throws IllegalStateException if the filter has been closed or already built
     */
    public void build() {
        checkAlive();
        if (built) {
            throw new IllegalStateException("Filter has already been built");
        }
        SketchOxideNative.ribbonfilter_build(nativePtr);
        built = true;
    }

    /**
     * Check if an item might be in the filter.
     *
     * Returns true if the item might be in the set (with false positive rate = fpr).
     * Returns false if the item is definitely not in the set.
     *
     * @param item the item to check (as raw bytes)
     * @return true if the item might be in the set, false if definitely not
     * @throws IllegalStateException if the filter has been closed or not yet built
     * @throws NullPointerException if item is null
     */
    public boolean contains(byte[] item) {
        checkAlive();
        if (!built) {
            throw new IllegalStateException("Must call build() before querying");
        }
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return SketchOxideNative.ribbonfilter_contains(nativePtr, item);
    }

    /**
     * Check if a string item might be in the filter.
     *
     * @param item the string to check
     * @return true if the item might be in the set, false if definitely not
     * @throws IllegalStateException if the filter has been closed or not yet built
     * @throws NullPointerException if item is null
     */
    public boolean contains(String item) {
        if (item == null) {
            throw new NullPointerException("item cannot be null");
        }
        return contains(item.getBytes());
    }

    /**
     * Check if the filter has been built.
     *
     * @return true if build() has been called, false otherwise
     */
    public boolean isBuilt() {
        return built;
    }

    /**
     * Merge another RibbonFilter into this one.
     *
     * <p><strong>Note:</strong> RibbonFilter does not support merge operations
     * due to its construction algorithm. To combine sets, create a new filter
     * with all items.
     *
     * @param other the filter to merge
     * @throws UnsupportedOperationException always
     */
    @Override
    public void merge(RibbonFilter other) {
        throw new UnsupportedOperationException(
                "RibbonFilter does not support merge operations. " +
                        "Create a new filter with combined items instead.");
    }

    /**
     * Serialize the filter to binary format.
     *
     * @return binary representation of the filter
     * @throws IllegalStateException if the filter has been closed or not yet built
     */
    public byte[] serialize() {
        checkAlive();
        if (!built) {
            throw new IllegalStateException("Must call build() before serializing");
        }
        return SketchOxideNative.ribbonfilter_serialize(nativePtr);
    }

    /**
     * Deserialize a RibbonFilter from binary format.
     *
     * The deserialized filter is already in built state and ready for queries.
     *
     * @param data the serialized data from {@link #serialize()}
     * @return a new RibbonFilter instance (already built)
     * @throws IllegalArgumentException if data is invalid or corrupt
     * @throws NullPointerException if data is null
     */
    public static RibbonFilter deserialize(byte[] data) {
        if (data == null) {
            throw new NullPointerException("data cannot be null");
        }

        long ptr = SketchOxideNative.ribbonfilter_deserialize(data);
        if (ptr == 0) {
            throw new IllegalArgumentException("Invalid or corrupt serialized RibbonFilter data");
        }

        // Deserialized filters are already built
        return new RibbonFilter(ptr, 100, 0.01, true);
    }

    /**
     * Get the expected number of items.
     *
     * @return the n parameter used at creation
     */
    public long getN() {
        return n;
    }

    /**
     * Get the target false positive rate.
     *
     * @return the fpr parameter used at creation
     */
    public double getFpr() {
        return fpr;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            freeNative();
            nativePtr = 0;
        }
    }

    @Override
    protected void freeNative() {
        if (nativePtr != 0) {
            SketchOxideNative.ribbonfilter_free(nativePtr);
        }
    }

    @Override
    public String toString() {
        if (nativePtr != 0) {
            return String.format("RibbonFilter(n=%d, fpr=%.6f, built=%s)",
                    n, fpr, built);
        }
        return "RibbonFilter(closed)";
    }
}

File: dotnet/SketchOxide/src/Membership/RibbonFilter.cs
using System;
using System.Text;
using SketchOxide.Native;

namespace SketchOxide.Membership;

/// <summary>
/// Ribbon Filter for space-efficient membership testing with two-phase construction.
///
/// A modern filter design that uses Ribbon (Rapid Incremental Boolean Banding ON the fly)
/// construction for excellent space efficiency. Elements are first inserted, then the
/// filter is finalized with a Build() call before queries can be made.
/// </summary>
/// <remarks>
/// <para>
/// The Ribbon filter provides:
/// - Excellent space efficiency (close to theoretical minimum)
/// - Two-phase construction: Insert elements, then Build to finalize
/// - Fast queries after construction
/// - No deletion support
/// - Configurable false positive rate
/// </para>
/// <para>
/// Usage pattern:
/// <code>
/// var filter = new RibbonFilter(1000, 0.01);
/// filter.Insert(data1);
/// filter.Insert(data2);
/// // ... insert all elements
/// bool success = filter.Build(); // Finalize the filter
/// if (success)
/// {
///     bool found = filter.Contains(query);
/// }
/// </code>
/// </para>
/// <para>
/// Ideal for applications where:
/// - Space efficiency is critical
/// - Two-phase construction is acceptable
/// - Filter is built once and queried many times
/// </para>
/// </remarks>
public sealed class RibbonFilter : NativeSketch
{
    private readonly ulong _size;
    private readonly double _fpr;
    private bool _isBuilt;

    /// <summary>
    /// Creates a new Ribbon filter with the specified expected number of elements and false positive rate.
    /// </summary>
    /// <param name="expectedElements">Expected number of elements to be inserted. Must be greater than 0.</param>
    /// <param name="falsePositiveRate">Desired false positive rate (e.g., 0.01 for 1%). Must be in range (0, 1).</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown if expectedElements is 0 or falsePositiveRate is outside (0, 1).</exception>
    /// <exception cref="OutOfMemoryException">Thrown if native memory allocation fails.</exception>
    public RibbonFilter(ulong expectedElements, double falsePositiveRate)
    {
        if (expectedElements == 0)
            throw new ArgumentOutOfRangeException(nameof(expectedElements), expectedElements, "Expected elements must be greater than 0");
        if (falsePositiveRate <= 0 || falsePositiveRate >= 1)
            throw new ArgumentOutOfRangeException(nameof(falsePositiveRate), falsePositiveRate, "False positive rate must be in range (0, 1)");

        _size = expectedElements;
        _fpr = falsePositiveRate;
        _isBuilt = false;
        NativePtr = SketchOxideNative.ribbonfilter_new(expectedElements, falsePositiveRate);

        if (NativePtr == 0)
            throw new OutOfMemoryException("Failed to allocate native RibbonFilter");
    }

    /// <summary>
    /// Private constructor for deserialization.
    /// </summary>
    private RibbonFilter(ulong size, double fpr, bool isBuilt, nuint ptr)
    {
        _size = size;
        _fpr = fpr;
        _isBuilt = isBuilt;
        NativePtr = ptr;
    }

    /// <summary>
    /// Gets the expected number of elements this filter was configured for.
    /// </summary>
    public ulong ExpectedElements
    {
        get
        {
            CheckAlive();
            return _size;
        }
    }

    /// <summary>
    /// Gets the target false positive rate.
    /// </summary>
    public double FalsePositiveRate
    {
        get
        {
            CheckAlive();
            return _fpr;
        }
    }

    /// <summary>
    /// Gets whether the filter has been built and is ready for queries.
    /// </summary>
    public bool IsBuilt
    {
        get
        {
            CheckAlive();
            return _isBuilt;
        }
    }

    /// <summary>
    /// Inserts an element into the filter during the construction phase.
    /// </summary>
    /// <param name="data">The bytes representing the element to insert.</param>
    /// <returns>
    /// <c>true</c> if the element was successfully inserted;
    /// <c>false</c> if insertion failed (e.g., filter already built or too many elements).
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    /// <exception cref="InvalidOperationException">Thrown if Build() has already been called.</exception>
    public bool Insert(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));
        if (_isBuilt) throw new InvalidOperationException("Cannot insert elements after Build() has been called");

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.ribbonfilter_insert(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Inserts a string element into the filter during the construction phase.
    /// </summary>
    /// <param name="value">The string to insert.</param>
    /// <returns>
    /// <c>true</c> if the element was successfully inserted;
    /// <c>false</c> if insertion failed.
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    /// <exception cref="InvalidOperationException">Thrown if Build() has already been called.</exception>
    public bool Insert(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Insert(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Finalizes the filter construction, making it ready for queries.
    /// </summary>
    /// <returns>
    /// <c>true</c> if the filter was successfully built;
    /// <c>false</c> if construction failed.
    /// </returns>
    /// <remarks>
    /// After calling Build(), no more elements can be inserted.
    /// The filter must be built before Contains() can be called.
    /// </remarks>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    /// <exception cref="InvalidOperationException">Thrown if Build() has already been called.</exception>
    public bool Build()
    {
        CheckAlive();
        if (_isBuilt) throw new InvalidOperationException("Filter has already been built");

        bool success = SketchOxideNative.ribbonfilter_build(NativePtr);
        if (success)
        {
            _isBuilt = true;
        }
        return success;
    }

    /// <summary>
    /// Tests whether an element might be in the set.
    /// </summary>
    /// <param name="data">The bytes representing the element to query.</param>
    /// <returns>
    /// <c>true</c> if the element might be in the set (may be a false positive);
    /// <c>false</c> if the element is definitely not in the set (no false negatives).
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    /// <exception cref="InvalidOperationException">Thrown if Build() has not been called yet.</exception>
    public bool Contains(ReadOnlySpan<byte> data)
    {
        CheckAlive();
        if (data == null) throw new ArgumentNullException(nameof(data));
        if (!_isBuilt) throw new InvalidOperationException("Filter must be built before querying. Call Build() first.");

        unsafe
        {
            fixed (byte* ptr = data)
            {
                return SketchOxideNative.ribbonfilter_contains(NativePtr, new Span<byte>(ptr, data.Length).ToArray(), (ulong)data.Length);
            }
        }
    }

    /// <summary>
    /// Tests whether a string element might be in the set.
    /// </summary>
    /// <param name="value">The string to query.</param>
    /// <returns>
    /// <c>true</c> if the element might be in the set (may be a false positive);
    /// <c>false</c> if the element is definitely not in the set (no false negatives).
    /// </returns>
    /// <exception cref="ArgumentNullException">Thrown if value is null.</exception>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    /// <exception cref="InvalidOperationException">Thrown if Build() has not been called yet.</exception>
    public bool Contains(string value)
    {
        if (value == null) throw new ArgumentNullException(nameof(value));
        return Contains(Encoding.UTF8.GetBytes(value));
    }

    /// <summary>
    /// Serializes the filter to a byte array.
    /// </summary>
    /// <returns>Serialized filter bytes.</returns>
    /// <exception cref="ObjectDisposedException">Thrown if the filter is disposed.</exception>
    public byte[] Serialize()
    {
        CheckAlive();
        return SketchOxideNative.ribbonfilter_serialize(NativePtr, out _);
    }

    /// <summary>
    /// Deserializes a Ribbon filter from a byte array.
    /// </summary>
    /// <param name="data">Serialized filter bytes.</param>
    /// <param name="expectedElements">The expected elements parameter used when creating the original filter.</param>
    /// <param name="falsePositiveRate">The false positive rate parameter used when creating the original filter.</param>
    /// <param name="wasBuilt">Whether the original filter had been built before serialization.</param>
    /// <returns>A new RibbonFilter instance.</returns>
    /// <exception cref="ArgumentNullException">Thrown if data is null.</exception>
    /// <exception cref="ArgumentException">Thrown if data is invalid.</exception>
    public static RibbonFilter Deserialize(byte[] data, ulong expectedElements, double falsePositiveRate, bool wasBuilt = true)
    {
        if (data == null) throw new ArgumentNullException(nameof(data));

        nuint ptr = SketchOxideNative.ribbonfilter_deserialize(data, (ulong)data.Length);
        if (ptr == 0)
            throw new ArgumentException("Failed to deserialize RibbonFilter: invalid data");

        return new RibbonFilter(expectedElements, falsePositiveRate, wasBuilt, ptr);
    }

    /// <summary>
    /// Returns a string representation of the filter.
    /// </summary>
    public override string ToString()
    {
        if (IsDisposed)
            return "RibbonFilter(disposed)";
        return $"RibbonFilter(expectedElements={_size}, fpr={_fpr:P2}, built={_isBuilt})";
    }

    /// <summary>
    /// Frees the native Rust instance.
    /// </summary>
    protected override void FreeNative()
    {
        if (NativePtr != 0)
        {
            SketchOxideNative.ribbonfilter_free(NativePtr);
            NativePtr = 0;
        }
    }
}
